File: ios/RadziTracker/Segmenter.swift
import Foundation

enum Segmenter {
  enum SegmentKind { case active, transit }

  struct TimelinePoint {
    let timestamp: Date
    let activity: SegmentActivity
    let speedKmh: Double?
  }

  enum SegmentActivity {
    case walking, running, cycling, automotive, stationary, unknown
    var isActive: Bool { self == .walking || self == .running || self == .cycling }
  }

  struct Segment {
    let start: Date
    let end: Date
    let kind: SegmentKind
    let activity: SegmentActivity
  }

  static let transitWindowSeconds: TimeInterval = 60
  static let transitSpeedThresholdKmh: Double = 25
  static let transitSpeedMinDurationSeconds: TimeInterval = 30

  /// Split a chronological timeline into active and transit segments.
  /// Transit: any continuous ≥60s run of automotive/unknown/stationary.
  /// For unknown activity: if sustained GPS speed > 25 km/h for ≥30s → treated as automotive.
  static func split(timeline: [TimelinePoint]) -> [Segment] {
    guard timeline.count >= 2 else { return [] }

    var segments: [Segment] = []
    var windowStart = timeline[0].timestamp
    var windowActivity = classifyActivity(timeline[0])
    var windowPoints: [TimelinePoint] = [timeline[0]]

    func flush(to endDate: Date) {
      guard windowStart < endDate else { return }
      let duration = endDate.timeIntervalSince(windowStart)
      let isTransit: Bool
      if windowActivity.isActive {
        isTransit = false
      } else {
        isTransit = duration >= transitWindowSeconds
      }
      // Check if unknown activity had sustained high speed for ≥30s (automotive proxy)
      let finalActivity: SegmentActivity
      if windowActivity == .unknown && windowPoints.count >= 2 {
        var sustainedHighSpeedSeconds = 0.0
        for i in 1..<windowPoints.count {
          let prev = windowPoints[i - 1]
          let curr = windowPoints[i]
          if (prev.speedKmh ?? 0) > transitSpeedThresholdKmh && (curr.speedKmh ?? 0) > transitSpeedThresholdKmh {
            sustainedHighSpeedSeconds += curr.timestamp.timeIntervalSince(prev.timestamp)
          }
        }
        finalActivity = sustainedHighSpeedSeconds >= transitSpeedMinDurationSeconds ? .automotive : .unknown
      } else {
        finalActivity = windowActivity
      }
      segments.append(Segment(start: windowStart, end: endDate,
                              kind: isTransit ? .transit : .active,
                              activity: finalActivity))
    }

    for i in 1..<timeline.count {
      let pt = timeline[i]
      let activity = classifyActivity(pt)
      if activity != windowActivity {
        flush(to: pt.timestamp)
        windowStart = pt.timestamp
        windowActivity = activity
        windowPoints = [pt]
      } else {
        windowPoints.append(pt)
      }
    }
    flush(to: timeline.last!.timestamp)
    return segments
  }

  private static func classifyActivity(_ pt: TimelinePoint) -> SegmentActivity { pt.activity }
}

File: android/app/src/main/java/com/radzi/app/tracking/Segmenter.kt
package com.radzi.app.tracking

// DORMANT — DATA COLLECTION ONLY. Ported from iOS Segmenter.swift for parity.
// Not consumed by live classification on Android (same as iOS).
object Segmenter {

    enum class SegmentKind { ACTIVE, TRANSIT }

    enum class SegmentActivity {
        WALKING, RUNNING, CYCLING, AUTOMOTIVE, STATIONARY, UNKNOWN;

        val isActive: Boolean get() = this == WALKING || this == RUNNING || this == CYCLING
    }

    data class TimelinePoint(
        val timestampMs: Long,
        val activity: SegmentActivity,
        val speedKmh: Double?
    )

    data class Segment(
        val startMs: Long,
        val endMs: Long,
        val kind: SegmentKind,
        val activity: SegmentActivity
    )

    private const val TRANSIT_WINDOW_MS = 60_000L
    private const val TRANSIT_SPEED_THRESHOLD_KMH = 25.0
    private const val TRANSIT_SPEED_MIN_DURATION_MS = 30_000L

    fun split(timeline: List<TimelinePoint>): List<Segment> {
        if (timeline.size < 2) return emptyList()

        val segments = mutableListOf<Segment>()
        var windowStart = timeline[0].timestampMs
        var windowActivity = timeline[0].activity
        var windowPoints = mutableListOf(timeline[0])

        fun flush(toMs: Long) {
            if (windowStart >= toMs) return
            val durationMs = toMs - windowStart
            val isTransit = if (windowActivity.isActive) {
                false
            } else {
                durationMs >= TRANSIT_WINDOW_MS
            }

            val finalActivity: SegmentActivity
            if (windowActivity == SegmentActivity.UNKNOWN && windowPoints.size >= 2) {
                var sustainedHighSpeedMs = 0L
                for (i in 1 until windowPoints.size) {
                    val prev = windowPoints[i - 1]
                    val curr = windowPoints[i]
                    if ((prev.speedKmh ?: 0.0) > TRANSIT_SPEED_THRESHOLD_KMH &&
                        (curr.speedKmh ?: 0.0) > TRANSIT_SPEED_THRESHOLD_KMH
                    ) {
                        sustainedHighSpeedMs += curr.timestampMs - prev.timestampMs
                    }
                }
                finalActivity = if (sustainedHighSpeedMs >= TRANSIT_SPEED_MIN_DURATION_MS)
                    SegmentActivity.AUTOMOTIVE else SegmentActivity.UNKNOWN
            } else {
                finalActivity = windowActivity
            }

            segments.add(
                Segment(windowStart, toMs, if (isTransit) SegmentKind.TRANSIT else SegmentKind.ACTIVE, finalActivity)
            )
        }

        for (i in 1 until timeline.size) {
            val pt = timeline[i]
            if (pt.activity != windowActivity) {
                flush(pt.timestampMs)
                windowStart = pt.timestampMs
                windowActivity = pt.activity
                windowPoints = mutableListOf(pt)
            } else {
                windowPoints.add(pt)
            }
        }
        flush(timeline.last().timestampMs)
        return segments
    }
}
